### PR TITLE
feature: add option to define output format

### LINKED ISSUE #248: output to *.md, but with an option svg or png
**Is your feature request related to a problem? Please describe.**
I have a lot of Markdown files and need to extract Mermaid text to images. Output to markdown works fine, but images are svg files. Confluence doesn't accept svg and I must use Confluence.

**Describe the solution you'd like**
An option for svg or png when output to md.

**Describe alternatives you've considered**
Now I run the command twice, `-o readme.md`, then `-o readme.png` and doing a `sed -i s/.svg/.png/ readme.md`

### PR DESCRIPTION
**Pull request by root360-AndreasUlm**
Signed-off-by: Andreas Ulm <andreas.ulm@skaylink.com>

## :bookmark_tabs: Summary

This commit adds the short option `-e` and the long option `--outputFormat` to overwrite the output format of images generated when handling markdown files.

Resolves #248 

## :straight_ruler: Design Decisions

The option defines a variable called outputFormat.
The variable is given to `run()` function and used instead of `path.extname(output).replace('.', '')` when calling `parseMMD()`.
The value is given via CLI argument or falls back to `path.extname(output).replace('.', ''),` and in case of markdown extension to 'svg' to keep current behaviour.

## :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid-cli/blob/master/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :bookmark: targeted `master` branch

### CODE REVIEW

**Comment by root360-AndreasUlm** [comment]
FYI: to run the test suite on my MacBook Pro M1 I had to increase the timeout by using:
```
yarn node --experimental-vm-modules $(yarn bin jest) --testTimeout 20000
```

File: src/index.js [hunk 1]
```diff
@@ -258,7 +271,7 @@ async function parseMMD (browser, definition, outputFormat, { viewport, backgrou
  * @param {boolean} [opts.quiet] - If set, suppress log output.
```
Comment (aloisklink) [inline src/index.js:271]: Assuming you move the default `outputFormat` logic into this `run()` function, we can use this doc string.

```suggestion
 * @param {boolean} [opts.quiet] - If set, suppress log output.
 * @param {"svg" | "png" | "pdf"} [opts.outputFormat] - Mermaid output format.
 * Defaults to `output` extension. Overrides `output` extension if set.
```

File: src/index.js [hunk 2]
```diff
@@ -107,6 +108,17 @@ async function cli () {
...
+  if (!outputFormat) {
+    outputFormat = path.extname(output).replace('.', '')
+  }
+  if (outputFormat === 'md') {
+    // fallback to svg in case no outputFormat is given and output file is MD
+    outputFormat = 'svg'
+  }
+  if (!/(?:svg|png|pdf)$/.test(outputFormat)) {
+    error('Output format must be one of "svg", "png" or "pdf"')
+  }
```
Comment (aloisklink) [inline src/index.js:120]: Can you move this into the `run()` function?

This library exposes the `run()` function, so we should move this default behavior into that function to maintain backwards compatibility (even if it's a bit slower).

We'd also need to change the error condition to `throw new Error('Output format must be one of "svg", "png" or "pdf"');`

**Review by aloisklink — CHANGES_REQUESTED** [review]
Great PR! Thanks for adding some great tests too!

Most of my review comments are minor.

The only important one is: can you move the default `outputFormat` checks out of the `cli()` function and into the `run()` function?
(we'll also need to change the `error()` call to instead `throw new Error(...)`)

That way we should be backwards compatible with anybody already using the `run()` function.

**Review by aloisklink — APPROVED** [review]
Awesome! Thanks for the changes. It all looks great to me, but I'll let @MindaugasLaganeckas do a final review before merging, in case there's anything I missed.

File: src-test/test.js [hunk 3]
```diff
@@ -228,13 +228,40 @@ describe("NodeJS API (import ... from '@mermaid-js/mermaid-cli')", () => {
...
+        ].map((file) => fs.rm(file, { force: true }))
+      )
+
+      await run(
+        'test-positive/mermaid.md', expectedOutputMd, { quiet: true, outputFormat: 'png' }
+      )
+
+      const markdownFile = await fs.readFile(expectedOutputMd, { encoding: 'utf8' })
+
+      // files should exist, and they should be SVGs
```
[the remark this answers is not in the document]
Comment (aloisklink) [reply src-test/test.js]: Fixed in https://github.com/mermaid-js/mermaid-cli/pull/382/commits/fa79772dedda70470830ad2af17b0e25b5c89a7a

File: src-test/test.js [hunk 4]
```diff
@@ -228,13 +228,40 @@ describe("NodeJS API (import ... from '@mermaid-js/mermaid-cli')", () => {
...
+      await Promise.all(expectedOutputPngs.map(async (expectedOutputPng) => {
+        // markdown file should point to png relative to md file
+        expect(markdownFile).toContain(relative('test-output', expectedOutputPng))
+
+        const pngFile = await fs.readFile(expectedOutputPng, { encoding: 'utf8' })
+        expect(pngFile).toMatch(/^.*PNG/)
```
[the remark this answers is not in the document]
Comment (aloisklink) [reply src-test/test.js]: Fixed in https://github.com/mermaid-js/mermaid-cli/pull/382/commits/fa79772dedda70470830ad2af17b0e25b5c89a7a

File: src/index.js (hunk 1 above)
Comment (aloisklink) [reply src/index.js:260]: Fixed in https://github.com/mermaid-js/mermaid-cli/pull/382/commits/fa79772dedda70470830ad2af17b0e25b5c89a7a

File: src/index.js (hunk 2 above)
Comment (aloisklink) [reply src/index.js]: Fixed in https://github.com/mermaid-js/mermaid-cli/pull/382/commits/fa79772dedda70470830ad2af17b0e25b5c89a7a and https://github.com/mermaid-js/mermaid-cli/pull/382/commits/c36859c2cf393e7b5d0cf8b5007c7bdee9e759ee

File: src/index.js [hunk 5]
```diff
@@ -281,7 +294,7 @@ async function run (input, output, { puppeteerConfig = {}, quiet = false, parseM
         //     I.e. if "out.png", use "out-1.png", "out-2.png", etc
         //   If it is an output `.md` file, use that to base .svg numbered diagrams on
         //     I.e. if "out.md". use "out-1.svg", "out-2.svg", etc
-        const outputFile = output.replace(/(\.(md|png|svg|pdf))$/, `-${diagrams.length + 1}$1`).replace(/(\.md)$/, '.svg')
+        const outputFile = output.replace(/(\.(md|png|svg|pdf))$/, `-${diagrams.length + 1}$1`).replace(/(\.md)$/, '.' + outputFormat)
```
[the remark this answers is not in the document]
Comment (aloisklink) [reply src/index.js]: Fixed in https://github.com/mermaid-js/mermaid-cli/pull/382/commits/fa79772dedda70470830ad2af17b0e25b5c89a7a

**Comment by MindaugasLaganeckas** [comment]
Thank you for a good PR @root360-AndreasUlm  and very thorough review @aloisklink 😄

**Comment by ValentinMi** [comment]
Exactly what i needed , thank you

**Pull request merged**